crates/tx5/src/ep3.rs

```diff
@@ -0,0 +1,609 @@
+        self.0.insert(id, until);
+    }
+
+    fn is_banned(&mut self, id: Id) -> bool {
+        let now = tokio::time::Instant::now();
+        if let Some(until) = self.0.get(&id).cloned() {
+            if now < until {
+                true
+            } else {
+                self.0.remove(&id);
+                false
+            }
+        } else {
+            false
+        }
+    }
+}
+
+pub(crate) struct EpShared {
+    config: Arc<Config3>,
+    this_id: Id,
+    ep_uniq: u64,
+    lair_tag: Arc<str>,
+    lair_client: LairClient,
+    sig_limit: Arc<tokio::sync::Semaphore>,
+    peer_limit: Arc<tokio::sync::Semaphore>,
+    recv_recon_limit: Arc<tokio::sync::Semaphore>,
+    weak_sig_map: Weak<Mutex<SigMap>>,
+    evt_send: EventSend<Ep3Event>,
+    ban_map: Mutex<BanMap>,
+    metric_conn_count:
+        influxive_otel_atomic_obs::AtomicObservableUpDownCounterI64,
+}
+
+/// Tx5 endpoint version 3.
+pub struct Ep3 {
+    ep: Arc<EpShared>,
+    _lair_keystore: lair_keystore_api::in_proc_keystore::InProcKeystore,
+    _sig_map: Arc<Mutex<SigMap>>,
+    listen_sigs: Arc<Mutex<Vec<tokio::task::JoinHandle<()>>>>,
+}
+
+impl std::fmt::Debug for Ep3 {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Ep3")
+            .field("this_id", &self.ep.this_id)
+            .field("ep_uniq", &self.ep.ep_uniq)
+            .finish()
+    }
+}
+
+impl Drop for Ep3 {
+    fn drop(&mut self) {
+        let handles = std::mem::take(&mut *self.listen_sigs.lock().unwrap());
+        for handle in handles {
+            handle.abort();
+        }
+    }
+}
+
+impl Ep3 {
+    /// Construct a new tx5 endpoint version 3.
+    pub async fn new(config: Arc<Config3>) -> (Self, EventRecv<Ep3Event>) {
+        use influxive_otel_atomic_obs::MeterExt;
+        use opentelemetry_api::metrics::MeterProvider;
+
+        let sig_limit = Arc::new(tokio::sync::Semaphore::new(
+            config.connection_count_max as usize,
+        ));
+
+        let peer_limit = Arc::new(tokio::sync::Semaphore::new(
+            config.connection_count_max as usize,
+        ));
+
+        let recv_recon_limit = Arc::new(tokio::sync::Semaphore::new(
+            config.incoming_message_bytes_max as usize,
+        ));
+
+        let lair_tag: Arc<str> =
+            rand_utf8::rand_utf8(&mut rand::thread_rng(), 32).into();
+
+        let passphrase = sodoken::BufRead::new_no_lock(
+            rand_utf8::rand_utf8(&mut rand::thread_rng(), 32).as_bytes(),
+        );
+
+        // this is a memory keystore,
+        // so weak persistence security is okay,
+        // since it will not be persisted.
+        // The private keys will still be mem_locked
+        // so they shouldn't be swapped to disk.
+        let keystore_config = PwHashLimits::Minimum
+            .with_exec(|| LairServerConfigInner::new("/", passphrase.clone()))
+            .await
+            .unwrap();
+
+        let _lair_keystore = PwHashLimits::Minimum
```

Comment:
This gets used, does it need to be `_` prefixed?

Reply:
It's not used once it goes into the struct, and I like:

```rust
Self {
    _lair_keystore,
```

better than:

```rust
Self {
    _lair_keystore: lair_keystore,
```